src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py

```diff
@@ -0,0 +1,379 @@
+
+_logger = logging.getLogger(__name__)
+
+
+class LargeQueryBreakdown:
+    r"""Optimization to break down large query plans into smaller partitions based on
+    estimated complexity score of the plan nodes.
+
+    This optimization works by analyzing computed query complexity score for each input
+    plan and breaking down the plan into smaller partitions if we detect valid node
+    candidates for partitioning. The partitioning is done by creating temp tables for the
+    partitioned nodes and replacing the partitioned subtree with the temp table selectable.
+
+    Example:
+        For a data pipeline with a large query plan created like so:
+
+            >>> base_df = session.sql("select 1 as A, 2 as B")
+            >>> df1 = base_df.with_column("A", F.col("A") + F.lit(1))
+            >>> x = 100
+            >>> for i in range(x):
+            >>>     df1 = df1.with_column("A", F.col("A") + F.lit(i))
+            >>> df1 = df1.group_by(F.col("A")).agg(F.sum(F.col("B")).alias("B"))
+
+            >>> df2 = base_df.with_column("B", F.col("B") + F.lit(1))
+            >>> for i in range(x):
+            >>>     df2 = df2.with_column("B", F.col("B") + F.lit(i))
+            >>> df2 = df2.group_by(F.col("B")).agg(F.sum(F.col("A")).alias("A"))
+
+            >>> union_df = df1.union_all(df2)
+            >>> final_df = union_df.with_column("A", F.col("A") + F.lit(1))
+
+        The corresponding query plan has the following structure:
+
+                                 projection on result
+                                           |
+                                       UNION ALL
+        Groupby + Agg (A) ---------------/   \------------------ Groupby + Agg (B)
+        with columns set 1                                      with columns set 2
+
+
+
+        Given the right complexity bounds, large query breakdown optimization will break down
+        the plan into smaller partition and give us the following plan:
+
+
+           Create Temp table (T1)                                    projection on result
+                    |                      ,                                   |
+            Groupby + Agg (A)                                              UNION ALL
+            with columns set 1                  Select * from T1 -----------/  \-----------  Groupby + Agg (B)
+                                                                                             with columns set 2
+    """
+
+    def __init__(
+        self,
+        session: Session,
+        query_generator: QueryGenerator,
+        logical_plans: List[LogicalPlan],
+    ) -> None:
+        self.session = session
+        self._query_generator = query_generator
+        self.logical_plans = logical_plans
+        self._parent_map = defaultdict(set)
+        self._tmp_plot_dir = os.path.join(tempfile.gettempdir(), "snowpark_plan_plots")
+
+    def apply(self) -> List[LogicalPlan]:
+        if is_active_transaction(self.session):
```

Comment:
then it means we have to eagerly issue this extra query every time evaluating a dataframe?

Reply:
It will get triggered when new compilation stage is enabled and large query breakdown is enabled